internal/proxy/proxy.go

```diff
@@ -54,44 +55,110 @@ type Config struct {
 // Client represents the state of the current instantiation of the proxy.
 type Client struct {
 	cmd    *cobra.Command
-	dialer *cloudsqlconn.Dialer
+	dialer DBDialer
 
 	// mnts is a list of all mounted sockets for this client
 	mnts []*socketMount
 }
 
-// NewClient completes the initial setup required to get the proxy to a "steady" state.
-func NewClient(ctx context.Context, cmd *cobra.Command, conf *Config) (*Client, error) {
-	d, err := cloudsqlconn.NewDialer(ctx)
-	if err != nil {
-		return nil, fmt.Errorf("error initializing dialer: %v", err)
+type portConfig struct {
+	global    int
+	postgres  int
+	mysql     int
+	sqlserver int
+}
+
+func newPortConfig(global int) *portConfig {
+	return &portConfig{
+		global:    global,
+		postgres:  5432,
+		mysql:     3306,
+		sqlserver: 1433,
+	}
+}
+
+// nextPort returns the next port based on the initial global value.
+func (c *portConfig) nextPort() int {
+	p := c.global
+	c.global++
+	return p
+}
+
+func (c *portConfig) nextDBPort(version string) int {
+	switch {
+	case strings.HasPrefix(version, "MYSQL"):
+		p := c.mysql
+		c.mysql++
+		return p
+	case strings.HasPrefix(version, "POSTGRES"):
+		p := c.postgres
+		c.postgres++
+		return p
+	case strings.HasPrefix(version, "SQLSERVER"):
+		p := c.sqlserver
+		c.sqlserver++
+		return p
+	default:
+		// Unexpected engine version, use global port setting instead.
+		return c.nextPort()
 	}
-	c := &Client{cmd: cmd, dialer: d}
+}
 
-	port := conf.Port
-	var inc int
+// DBDialer is a dialer for a Cloud SQL instance.
+type DBDialer interface {
+	// Dial returns a connection to the specified instance.
+	Dial(ctx context.Context, inst string, opts ...cloudsqlconn.DialOption) (net.Conn, error)
+	// EngineVersion retrieves the provided instance's database version (e.g.,
+	// POSTGRES_14)
+	EngineVersion(ctx context.Context, inst string) (string, error)
+	// Close terminates all background operations and stops the dialer.
+	Close() error
+}
+
+// NewClient completes the initial setup required to get the proxy to a "steady" state.
+func NewClient(ctx context.Context, d DBDialer, cmd *cobra.Command, conf *Config) (*Client, error) {
+	var mnts []*socketMount
+	pc := newPortConfig(conf.Port)
+	var wg sync.WaitGroup
+	for _, inst := range conf.Instances {
+		wg.Add(1)
+		go func(i InstanceConnConfig) {
+			// Initiate refresh operation
+			d.EngineVersion(ctx, inst.Name)
+			wg.Done()
+		}(inst)
+	}
+	wg.Wait()
```

Comment:
I _think_ (unless I'm missing something) that the waitgroup is overkill - calls to EngineVersion will block until the refresh is done, so I don't think we need to wait for them all to complete before referencing.

Reply:
Its only merit that I can think of is to communicate intention (i.e., when mounting the sockets, we want immediate return values from the call). But thinking about the performance, I agree the wait group doesn't do anything better.

Reply:
I think in that case, it's better to just leave a comment explaining the intent rather than rely on the wg here.

Reply:
The wait group just means the waiting is consolidated to `wg.Wait`. Whereas without it, we will spend some time waiting for `d.EngineVersion` to return on its second invocation.